pkg/issuer/acme/setup.go

```diff
@@ -37,35 +37,55 @@ import (
 	errorAccountVerificationFailed = "ErrVerifyACMEAccount"
 	errorAccountUpdateFailed       = "ErrUpdateACMEAccount"
+	errorInvalidConfig             = "InvalidConfig"
+	errorInvalidURL                = "InvalidURL"
 
 	successAccountRegistered = "ACMEAccountRegistered"
 	successAccountVerified   = "ACMEAccountVerified"
 
-	messageAccountRegistrationFailed = "Failed to register ACME account: "
-	messageAccountVerificationFailed = "Failed to verify ACME account: "
-	messageAccountUpdateFailed       = "Failed to update ACME account:"
-	messageAccountRegistered         = "The ACME account was registered with the ACME server"
-	messageAccountVerified           = "The ACME account was verified with the ACME server"
+	messageAccountRegistrationFailed     = "Failed to register ACME account: "
+	messageAccountVerificationFailed     = "Failed to verify ACME account: "
+	messageAccountUpdateFailed           = "Failed to update ACME account:"
+	messageAccountRegistered             = "The ACME account was registered with the ACME server"
+	messageAccountVerified               = "The ACME account was verified with the ACME server"
+	messageNoSecretKeyGenerationDisabled = "the ACME issuer config has 'disableAccountKeyGeneration' set to true, but the secret was not found: "
+	messageInvalidPrivateKey             = "Account private key is invalid: "
+
+	messageTemplateUpdateToV2              = "Your ACME server URL is set to a v1 endpoint (%s). You should update the spec.acme.server field to %q"
+	messageTemplateNotRSA                  = "ACME private key in %q is not of type RSA"
+	messageTemplateFailedToParseURL        = "Failed to parse existing ACME server URI %q: %v"
+	messageTemplateFailedToParseAccountURL = "Failed to parse existing ACME account URI %q: %v"
 )
 
 // Setup will verify an existing ACME registration, or create one if not
 // already registered.
 func (a *Acme) Setup(ctx context.Context) error {
 	log := logf.FromContext(ctx)
 
+	// Correct reason and message for issuer's Ready condition must be always set
+	// before returning from this function. Status must be set if not false.
+	status := cmmeta.ConditionFalse
+	var reason, msg string
+	defer func() {
+		apiutil.SetIssuerCondition(a.issuer,
+			a.issuer.GetGeneration(),
+			v1.IssuerConditionReady,
+			status,
+			reason,
+			msg)
+	}()
+
 	// check if user has specified a v1 account URL, and set a status condition if so.
 	if newURL, ok := acmev1ToV2Mappings[a.issuer.GetSpec().ACME.Server]; ok {
-		apiutil.SetIssuerCondition(a.issuer, a.issuer.GetGeneration(), v1.IssuerConditionReady, cmmeta.ConditionFalse, "InvalidConfig",
-			fmt.Sprintf("Your ACME server URL is set to a v1 endpoint (%s). "+
-				"You should update the spec.acme.server field to %q", a.issuer.GetSpec().ACME.Server, newURL))
+		reason = errorInvalidConfig
+		msg = fmt.Sprintf(messageTemplateUpdateToV2, a.issuer.GetSpec().ACME.Server, newURL)
 		// return nil so that Setup only gets called again after the spec is updated
```

Comment:
The parent controller could potentially tell that there is no point to attempt to set up a failed issuer again if:

1. Issuer is in not `Ready` state with Reason `InvalidConfig` or `errorInvalidURL`
2. The `issuer.status.conditions.0.observedGeneration` matches `issuer.metadata.observedGeneration`

If we fail with `InvalidConfig` or `errorInvalidURL` , there won't have been any API calls made to ACME yet at a point as far as I can tell. We would save some calls to Kubernetes API server though.

It seems like all the other failure cases involve some external state, so it would not be possible to tell if we should retry from the issuer spec alone. 
Perhaps we could implement a backoff, similar to how it's done for Certificates with [shouldBackoffReissuingOnFailure](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/trigger/trigger_controller.go#L166)?

Reply:
As discussed at standup, I think we should extend the validation in https://github.com/jetstack/cert-manager/blob/0ff2b8778c51e6cebe140a6b196e7a9a28cbee87/pkg/internal/apis/certmanager/validation/issuer.go#L106-L134 and then call `ValidateIssuerSpec` in the partent controller, before passing the Issuer to the `Setup` function...that way the Setup implementations can be guaranteed that the supplied GenericIssuer is valid and need not perform these checks.

But that's something for a future PR.